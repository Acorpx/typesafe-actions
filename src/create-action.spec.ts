import { createAction, getType } from './';

describe('createAction', () => {
  it('with type only', () => {
    const increment = createAction('WITH_TYPE_ONLY');
    const typeLiteral: 'WITH_TYPE_ONLY' = getType(increment);
    expect(typeLiteral).toBe('WITH_TYPE_ONLY');
  });

  it('with type only using symbol nominal-type pattern', () => {
    enum Increment {}
    const INCREMENT = (Symbol(1) as any) as Increment & string;
    const increment = createAction(INCREMENT, resolve => {
      return () => resolve();
    });
    const typeLiteral: typeof INCREMENT = getType(increment);
    expect(typeLiteral).toBe(INCREMENT);
    expect(typeLiteral).not.toBe('WITH_TYPE_ONLY');
  });

  it('with payload', () => {
    const add = createAction('WITH_PAYLOAD', resolve => {
      return (amount: number) => resolve(amount);
    });
    const typeLiteral: 'WITH_PAYLOAD' = getType(add);
    expect(typeLiteral).toBe('WITH_PAYLOAD');
  });

  it('with payload and meta', () => {
    const getTodo = createAction('GET_TODO', resolve => {
      return (id: string, token: string) => resolve(id, token);
    });
    const typeLiteral: 'GET_TODO' = getType(getTodo);
    expect(typeLiteral).toBe('GET_TODO');
  });
});
