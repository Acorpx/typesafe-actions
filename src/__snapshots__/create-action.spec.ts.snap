// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux Actions - with meta - obj param withMeta({ meta: 'info message' }) (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string>"`;

exports[`Redux Actions - with meta - obj union param withMeta({ meta: 'info message' }) (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string | number | null>"`;

exports[`Redux Actions - with meta - obj union param withMeta({ meta: 3 }) (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string | number | null>"`;

exports[`Redux Actions - with meta - obj union param withMeta({ meta: null }) (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string | number | null>"`;

exports[`Redux Actions - with payload - obj param withPayload({ payload: 'info message' }) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string>"`;

exports[`Redux Actions - with payload - obj union param withPayload({ payload: 'info message' }) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`Redux Actions - with payload - obj union param withPayload({ payload: 3 }) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`Redux Actions - with payload - obj union param withPayload({ payload: null }) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`Redux Actions - with payload and meta - multiple primitive params withPayloadAndMeta('Piotr', 'Hello!') (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", string, { username: string; message: string; }>"`;

exports[`Redux Actions - with payload and meta - obj param withPayloadAndMeta({ username: 'Piotr', message: 'Hello!' }) (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", string, { username: string; message: string; }>"`;

exports[`toString() method return a type actionCreator.toString() (type) should match snapshot 1`] = `"string"`;

exports[`type-safe usage with higher-order function setUserSetting('settingA', 'foo') (type) should match snapshot 1`] = `"TH.PayloadAction<\\"SET_USER_SETTING\\", { setting: \\"settingA\\"; newValue: string; }>"`;

exports[`type-safe usage with higher-order function setUserSetting('settingA', 0) (type) should match snapshot 1`] = `"Argument of type '0' is not assignable to parameter of type 'string'."`;

exports[`type-safe usage with higher-order function setUserSetting('settingB', 'foo') (type) should match snapshot 1`] = `"Argument of type '\\"foo\\"' is not assignable to parameter of type 'number'."`;

exports[`type-safe usage with higher-order function setUserSetting('settingB', 0) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"SET_USER_SETTING\\", { setting: \\"settingB\\"; newValue: number; }>"`;

exports[`with meta - any withMeta(undefined, 'token') (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, any>"`;

exports[`with meta withMeta(undefined, 'token') (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string>"`;

exports[`with payload - any withPayload(10) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", any>"`;

exports[`with payload - boolean withPayload(true) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", boolean>"`;

exports[`with payload - literal string union withPayload('up') (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", \\"up\\" | \\"down\\" | \\"unknown\\">"`;

exports[`with payload - number withPayload(10) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", number>"`;

exports[`with payload - primitives union withPayload('foo') (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`with payload - primitives union withPayload(3) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`with payload - primitives union withPayload(null) (type) should match snapshot 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`with payload and meta withPayloadAndMeta(1, 'token') (type) should match snapshot 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", number, string>"`;

exports[`with symbol withSymbol() (type) should match snapshot 1`] = `"TH.EmptyAction<any>"`;

exports[`with type only - undefined withTypeOnly() (type) should match snapshot 1`] = `"TH.EmptyAction<\\"CREATE_ACTION\\">"`;

exports[`with type only withTypeOnly() (type) should match snapshot 1`] = `"TH.EmptyAction<\\"CREATE_ACTION\\">"`;
